packages/mui-joy/src/Autocomplete/Autocomplete.tsx

```diff
@@ -0,0 +1,890 @@
+import Chip, { chipClasses } from '../Chip';
+import ChipDelete from '../ChipDelete';
+import { IconButtonOwnerState } from '../IconButton';
+import { inputClasses } from '../Input';
+import List from '../List';
+import ListProvider from '../List/ListProvider';
+import ListSubheader from '../ListSubheader';
+import ListItem from '../ListItem';
+import autocompleteClasses, { getAutocompleteUtilityClass } from './autocompleteClasses';
+import {
+  AutocompleteProps,
+  AutocompleteRenderGroupParams,
+  AutocompleteRenderGetTagProps,
+  AutocompleteOwnerState,
+} from './AutocompleteProps';
+import FormControlContext from '../FormControl/FormControlContext';
+import { AutocompleteOptionRoot } from '../AutocompleteOption/AutocompleteOption';
+import { AutocompleteListboxRoot } from '../AutocompleteListbox/AutocompleteListbox';
+import useSlot from '../utils/useSlot';
+
+type OwnerState = Omit<AutocompleteOwnerState<any, any, any, any>, 'onChange' | 'defaultValue'>;
+
+const defaultIsActiveElementInListbox = (listboxRef: React.RefObject<HTMLElement>) =>
+  listboxRef.current !== null && listboxRef.current.contains(document.activeElement);
+const defaultGetOptionLabel = <T extends unknown>(option: T) =>
+  (option as { label: string }).label ?? option;
+const defaultLimitTagsText = (more: string | number) => `+${more}`;
+const defaultRenderGroup = (params: AutocompleteRenderGroupParams) => (
+  <ListItem key={params.key} nested>
+    <ListSubheader sticky>{params.group}</ListSubheader>
+    <List>{params.children}</List>
+  </ListItem>
+);
+
+const useUtilityClasses = (ownerState: OwnerState) => {
+  const { focused, fullWidth, hasClearIcon, hasPopupIcon, popupOpen } = ownerState;
+
+  const slots = {
+    root: [
+      'root',
+      focused && 'focused',
+      fullWidth && 'fullWidth',
+      hasClearIcon && 'hasClearIcon',
+      hasPopupIcon && 'hasPopupIcon',
+    ],
+    clearIndicator: ['clearIndicator'],
+    popupIndicator: ['popupIndicator', popupOpen && 'popupIndicatorOpen'],
+    listbox: ['listbox'],
+    option: ['option'],
+    loading: ['loading'],
+    noOptions: ['noOptions'],
+    limitTag: ['limitTag'],
+  };
+
+  return composeClasses(slots, getAutocompleteUtilityClass, {});
+};
+
+const AutocompleteRoot = styled('div', {
+  name: 'JoyAutocomplete',
+  slot: 'Root',
+  overridesResolver: (props, styles) => styles.root,
+})<{ ownerState: OwnerState }>(({ ownerState }) => {
+  let endDecoratorCount = 0;
+  if (ownerState.hasClearIcon) {
+    endDecoratorCount += 1;
+  }
+  if (ownerState.hasPopupIcon) {
+    endDecoratorCount += 1;
+  }
+  return [
+    {
+      '--Autocomplete-endDecorator-count': endDecoratorCount,
+      ...(ownerState.fullWidth && {
+        width: '100%',
+      }),
+      /* Avoid double tap issue on iOS */
+      '@media (pointer: fine)': {
+        [`&:hover .${autocompleteClasses.clearIndicator}`]: {
+          visibility: 'visible',
+        },
+      },
+      [`& .${inputClasses.root}`]: {
```

Comment:
I believe we should have the `inputClasses` as a prop, so that if someone uses different input component, they can specify the classes. Or we can even have an `AutocompleteInput` that people can extend.

Reply:
> we should have the inputClasses as a prop

Maybe not if you look at the styles inside the `& .${inputClasses.root}`, it is specific to override Joy UI `Input` component.

Meaning if you have a custom input, it is better to not have the style affect your component. If you really do, it is simply this:

```js
// where `inputClasses` is from '@mui/material/Input';
<CustomInput className={inputClasses.root}>
```

> Or we can even have an AutocompleteInput that people can extend

I went with that path before but it is more complicated and without much value. Only the `inputClasses.root` specificity can be removed but we still have other overrides, e.g. `inputClasses.input`, `inputClasses.decorator`.

However, the extension experience brings no difference:

```js
const CustomAutocompleteInput = styled(Input)({
...
});
const CustomAutocompleteInput = styled(AutocompleteInput)({
...
})

<Autocomplete renderInput={() => <CustomAutocompleteInput />}>
```

So I rather go with using `Input` component to get similar experience to Material UI.

Reply:
Ok, let's keep it simple at start and see how it will go